docs/guide/data-security.asciidoc

```diff
@@ -0,0 +1,410 @@
+
+**Agent configuration options**
+
+* (Node.js) Remove errors raised by the server-side process:
+Disable with {apm-node-ref-v}/configuration.html#capture-exceptions[captureExceptions].
+
+* (Java) Remove process arguments from transactions:
+* Disabled by default with {apm-java-ref-v}/config-reporter.html#config-include-process-args[`include_process_args`].
+
+[discrete]
+[[custom-filters]]
+=== Custom filters
+
+There are two ways to filter or obfuscate other types APM data:
+
+* <<filter-ingest-pipeline>>
+* <<filter-in-agent>>
+
+[discrete]
+[[filter-ingest-pipeline]]
+==== Create an ingest node pipeline filter
+
+Ingest node pipelines specify a series of processors that transform data in a specific way.
+Transformation happens prior to indexing–inflicting no performance overhead on the monitored application.
+Pipelines are a flexible and easy way to filter or obfuscate Elastic APM data.
+
+**Example**
+
+Say you decide to <<filter-http-body,enable the capturing of HTTP request bodies>>,
+but quickly notice that sensitive information is being collected in the
+`http.request.body.original` field:
+
+[source,json]
+----
+{
+  "email": "test@abc.com",
+  "password": "hunter2"
+}
+----
+
+To obfuscate the passwords stored in the request body,
+you can use a series of {ref}/processors.html[ingest processors].
+To start, create a pipeline with a simple description and an empty array of processors:
+
+[source,json]
+----
+{
+  "pipeline": {
+    "description": "redact http.request.body.original.password",
+    "processors": [] <1>
+  }
+}
+----
+<1> The processors defined below will go in this array
+
+Add the first processor to the processors array.
+Because the agent captures the request body as a string, you must use the
+{ref}/json-processor.html[JSON processor] to convert the original field value into a structured JSON object.
+Save this JSON object in a new field:
+
+[source,json]
+----
+{
+  "json": {
+    "field": "http.request.body.original",
+    "target_field": "http.request.body.original_json",
+    "ignore_failure": true
+  }
+}
+----
+
+Next, use the {ref}/remove-processor.html[remove processor] to remove the `body.original` field
+if the new `body.original_json` field was created:
+
+[source,json]
+----
+{
+  "remove": {
+    "field": "http.request.body.original",
+    "if": "ctx?.http?.request?.body?.original_json != null",
+    "ignore_failure": true
+  }
+}
+----
+
+Finally, {ref}/set-processor.html[set] the value of the new `body.original_json` field to `"redacted"`–again,
+only if the new `body.original_json` field was created:
+
+[source,json]
+----
+{
+  "set": {
+    "field": "http.request.body.original_json.password",
+    "value": "redacted",
+    "if": "ctx?.http?.request?.body?.original_json != null"
+  }
+}
+----
+
+With the pipeline defined,
+use the {ref}/put-pipeline-api.html[create or update pipeline API] to register the new pipeline in {es}.
+Name the pipeline `apm_redacted_body_password`:
+
```

Comment:
We might want to add another optional processor that converts `original_json` back to a stringified representation in `http.request.body.original` and remove `http.request.body.original_json`.

Reply:
I've updated the pipeline example in 562c87c. I spent some time digging through the [processor reference](https://www.elastic.co/guide/en/elasticsearch/reference/master/processors.html), but as far as I can tell, there's no easy way to stringify JSON with a processor–am I missing something obvious? The new example does the following:

1. `http.request.body.original` --> JSON --> `http.request.body.original_json`.
2. Sets `http.request.body.original_json.password` to `"redacted"`.
3. Sets `http.request.body.original` to `http.request.body.original_json`.
4. Removes `http.request.body.original_json`.

Happy to update again if needed.

Reply:
> there's no easy way to stringify JSON with a processor

hmm, bummer

> Sets http.request.body.original to http.request.body.original_json

I think that would cause a mapping conflict: `http.request.body.original` has a mapping type of `string` but you'd be setting an `object` value.

Reply:
Oof, good point. Fixed in 89988a8 with the convert processor. This converts from JSON back to a string when setting the value of `body.original_json` to `body.original`